subprojects/build-option/src/main/java/org/gradle/internal/buildoption/BuildOption.java

```diff
@@ -30,18 +31,46 @@
  */
 public interface BuildOption<T> {
 
-    @Nullable String getGradleProperty();
+    @Nullable
+    String getGradleProperty();
+
     void applyFromProperty(Map<String, String> properties, T settings);
+
     void configure(CommandLineParser parser);
+
     void applyFromCommandLine(ParsedCommandLine options, T settings);
 
     enum Origin {
-        GRADLE_PROPERTY, COMMAND_LINE;
+        GRADLE_PROPERTY {
+            @Override
+            public void handleInvalidValue(BuildOption<?> option, String value, String hint) {
+                String message = String.format("Value '%s' given for %s Gradle property is invalid%s.", value, option.getGradleProperty(), hintMessage(hint));
+                throw new IllegalArgumentException(message);
+            }
+        },
+        COMMAND_LINE {
+            @Override
+            public void handleInvalidValue(BuildOption<?> option, String value, String hint) {
+                if (!(option instanceof StringBuildOption)) {
```

Comment:
`instanceof` is ugly here. @bmbaisch Do you have better ideas?

Reply:
:x: I think I'd personally just use the method signature `void handleInvalidValue(String source, String value, String hint)`. The `BuildOption` parameter doesn't quite work for command line options as you don't know exactly which option was declared by the user on the command line. It's more of a guess. Plus you'll also have to do the special type handling you are doing here. If you just pass in a String then the calling code can decide what option was actually used.